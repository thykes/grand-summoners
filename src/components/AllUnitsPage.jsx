// src/components/AllUnitsPage.jsx
import React, { useState, useMemo } from 'react';
import { useNavigate } from 'react-router-dom';
import PageHeader from './PageHeader';
import CompactUnitCard from './CompactUnitCard'; // Import the new component
import { getArtGenInfo, artsGenData } from '../data/artGenGroups';

export default function AllUnitsPage({ allUnits, myUnits, onToggleUnit, setSelectedUnit, showMyUnitsFilter, setShowMyUnitsFilter, onClearAll }) {
  const navigate = useNavigate();
  const [filters, setFilters] = useState({ element: 'All', archetype: 'All', artGenGroup: 'ALL_UNITS', searchQuery: '' });
  const [sortOrder, setSortOrder] = useState('release');

  const orderedArtGenGroups = useMemo(() => {
    return Array.from(new Set(Object.values(artsGenData).map(info => info.type))).sort();
  }, []);
  
  const filterOptions = useMemo(() => ({
    primaryArchetypes: [...new Set(allUnits.map(u => u.Primary_Archetype).filter(Boolean))].sort(),
  }), [allUnits]);

  const filteredAndSortedUnits = useMemo(() => {
    let processedUnits = showMyUnitsFilter ? allUnits.filter(u => myUnits.has(u.id)) : [...allUnits];
    processedUnits = processedUnits.filter(unit => {
      const artGenInfo = getArtGenInfo(unit.Unit_Name);
      let artGenMatch;
      switch (filters.artGenGroup) {
        case 'ALL_UNITS': artGenMatch = true; break;
        case 'NONE': artGenMatch = !artGenInfo; break;
        case 'ANY_ART_GEN': artGenMatch = !!artGenInfo; break;
        default: artGenMatch = artGenInfo?.type === filters.artGenGroup; break;
      }
      return artGenMatch &&
        (filters.element === 'All' || unit.Element === filters.element) &&
        (filters.archetype === 'All' || unit.Primary_Archetype === filters.archetype) &&
        (unit.Unit_Name.toLowerCase().includes(filters.searchQuery.toLowerCase()));
    });
    
    processedUnits.sort((a, b) => {
      if (sortOrder === 'A-Z') return a.Unit_Name.localeCompare(b.Unit_Name);
      if (sortOrder === 'Z-A') return b.Unit_Name.localeCompare(a.Unit_Name);
      return (a.Release_Order_Index || 99999) - (b.Release_Order_Index || 99999);
    });

    return processedUnits;
  }, [allUnits, filters, sortOrder, myUnits, showMyUnitsFilter]);
  
  const handleFilterChange = (e) => setFilters(prev => ({ ...prev, [e.target.name]: e.target.value }));
  const handleUnitClick = (unit) => {
    setSelectedUnit(unit);
    navigate(`/unit/${unit.id}`);
  };
  const handleClearAll = () => {
    if (window.confirm("Are you sure you want to remove all units from your collection?")) onClearAll();
  };

  return (
    <div className="w-full">
      <PageHeader title="Unit Database" subtitle="Browse all units and manage your collection." />
      <main className="w-full">
        <div className="card mb-8">
          <input type="text" name="searchQuery" placeholder="Search by unit name..." value={filters.searchQuery} onChange={handleFilterChange} className="bg-gray-700/80 text-white rounded-md p-3 w-full border border-gray-600 focus:ring-2 focus:ring-cyan-500 mb-4" />
          <div className="grid grid-cols-2 md:grid-cols-3 lg:grid-cols-5 gap-4">
            <select name="element" onChange={handleFilterChange} value={filters.element} className="bg-gray-700/80 text-white rounded-md p-2 w-full border border-gray-600">
              <option value="All">All Elements</option>
              <option value="Fire">Fire</option><option value="Water">Water</option><option value="Earth">Earth</option><option value="Light">Light</option><option value="Dark">Dark</option>
            </select>
            <select name="archetype" onChange={handleFilterChange} value={filters.archetype} className="bg-gray-700/80 text-white rounded-md p-2 w-full border border-gray-600">
              <option value="All">All Archetypes</option>
              {filterOptions.primaryArchetypes.map(type => <option key={type} value={type}>{type}</option>)}
            </select>
            <select name="artGenGroup" onChange={handleFilterChange} value={filters.artGenGroup} className="bg-gray-700/80 text-white rounded-md p-2 w-full border border-gray-600">
              <option value="ALL_UNITS">All Art Gen Groups</option>
              <option value="NONE">None</option>
              {orderedArtGenGroups.map(group => <option key={group} value={group}>Type {group}</option>)}
              <option value="ANY_ART_GEN">Any Art Gen Unit</option>
            </select>
            <select onChange={(e) => setSortOrder(e.target.value)} value={sortOrder} className="bg-gray-700/80 text-white rounded-md p-2 w-full border border-gray-600">
              <option value="release">Sort: Release Order</option>
              <option value="A-Z">Sort: Name (A-Z)</option> 
              <option value="Z-A">Sort: Name (Z-A)</option> 
            </select>
            <div className="flex items-center justify-center p-2 bg-gray-900/50 rounded-md border border-gray-700">
              <label htmlFor="my-units-toggle" className="mr-3 text-sm font-bold text-gray-300 select-none">Show My Units</label>
              <input type="checkbox" id="my-units-toggle" checked={showMyUnitsFilter} onChange={(e) => setShowMyUnitsFilter(e.target.checked)} className="form-checkbox h-5 w-5"/>
            </div>
          </div>
          <div className="mt-4 text-center">
            <button onClick={handleClearAll} className="text-sm font-semibold text-red-400 hover:text-red-300">Clear All My Units</button>
          </div>
        </div>
        <div className="grid grid-cols-4 sm:grid-cols-6 md:grid-cols-8 lg:grid-cols-10 xl:grid-cols-12 gap-x-4 gap-y-8">
          {filteredAndSortedUnits.map(unit => (
            <CompactUnitCard 
              unit={unit} 
              key={unit.id} 
              onUnitClick={handleUnitClick}
              isSelected={myUnits.has(unit.id)}
              onToggleSelect={onToggleUnit} // <-- Pass the correct prop
            />
          ))}
        </div>
      </main>
    </div>
  );
}