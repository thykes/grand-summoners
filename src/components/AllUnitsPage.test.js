// src/components/AllUnitsPage.test.js
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import AllUnitsPage from './AllUnitsPage';

describe('AllUnitsPage', () => {
  test('renders the header', () => {
    render(
      <MemoryRouter>
        <AllUnitsPage
          allUnits={[]}
          myUnits={new Set()}
          onToggleUnit={jest.fn()}
          setSelectedUnit={jest.fn()}
          showMyUnitsFilter={false}
          setShowMyUnitsFilter={jest.fn()}
          onClearAll={jest.fn()}
        />
      </MemoryRouter>
    );

    expect(screen.getByText(/Unit Database/i)).toBeInTheDocument();
  });

  test('renders units safely even with missing data', () => {
    const brokenUnits = [
      { id: '1' }, // missing Unit_Name
      { id: '2', Unit_Name: 'Hero X', Element: 'Fire' },
    ];

    render(
      <MemoryRouter>
        <AllUnitsPage
          allUnits={brokenUnits}
          myUnits={new Set(['2'])}
          onToggleUnit={jest.fn()}
          setSelectedUnit={jest.fn()}
          showMyUnitsFilter={false}
          setShowMyUnitsFilter={jest.fn()}
          onClearAll={jest.fn()}
        />
      </MemoryRouter>
    );

    // It should render Hero X but not crash on the broken unit
    expect(screen.getByText(/Hero X/i)).toBeInTheDocument();
  });

  test('respects "Show My Units" filter', () => {
    const units = [
      { id: '1', Unit_Name: 'Hero Y', Element: 'Water' },
      { id: '2', Unit_Name: 'Hero Z', Element: 'Earth' },
    ];

    render(
      <MemoryRouter>
        <AllUnitsPage
          allUnits={units}
          myUnits={new Set(['2'])}
          onToggleUnit={jest.fn()}
          setSelectedUnit={jest.fn()}
          showMyUnitsFilter={true}
          setShowMyUnitsFilter={jest.fn()}
          onClearAll={jest.fn()}
        />
      </MemoryRouter>
    );

    // Only Hero Z should appear since filter is ON
    expect(screen.getByText(/Hero Z/i)).toBeInTheDocument();
    expect(screen.queryByText(/Hero Y/i)).not.toBeInTheDocument();
  });

  test('matches snapshot with sample data', () => {
    const units = [
      { id: '1', Unit_Name: 'Hero Alpha', Element: 'Fire' },
      { id: '2', Unit_Name: 'Hero Beta', Element: 'Water' },
    ];

    const { asFragment } = render(
      <MemoryRouter>
        <AllUnitsPage
          allUnits={units}
          myUnits={new Set(['1'])}
          onToggleUnit={jest.fn()}
          setSelectedUnit={jest.fn()}
          showMyUnitsFilter={false}
          setShowMyUnitsFilter={jest.fn()}
          onClearAll={jest.fn()}
        />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
