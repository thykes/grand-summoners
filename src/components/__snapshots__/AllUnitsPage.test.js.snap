// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllUnitsPage matches snapshot with sample data 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <header
      class="text-center mb-10 w-full"
    >
      <h1
        class="text-4xl sm:text-5xl lg:text-6xl font-extrabold text-transparent bg-clip-text  bg-gradient-to-r from-cyan-300 to-fuchsia-500 mb-2 drop-shadow-lg"
      >
        Unit Database
      </h1>
      <h2
        class="text-xl font-semibold text-gray-400"
      >
        Browse all units and manage your collection.
      </h2>
    </header>
    <main
      class="w-full"
    >
      <div
        class="card mb-8"
      >
        <input
          class="bg-gray-700/80 text-white rounded-md p-3 w-full border border-gray-600 focus:ring-2 focus:ring-cyan-500 mb-4"
          name="searchQuery"
          placeholder="Search by unit name..."
          type="text"
          value=""
        />
        <div
          class="grid grid-cols-2 md:grid-cols-3 lg:grid-cols-5 gap-4"
        >
          <select
            class="bg-gray-700/80 text-white rounded-md p-2 w-full border border-gray-600"
            name="element"
          >
            <option
              value="All"
            >
              All Elements
            </option>
            <option
              value="Fire"
            >
              Fire
            </option>
            <option
              value="Water"
            >
              Water
            </option>
            <option
              value="Earth"
            >
              Earth
            </option>
            <option
              value="Light"
            >
              Light
            </option>
            <option
              value="Dark"
            >
              Dark
            </option>
          </select>
          <select
            class="bg-gray-700/80 text-white rounded-md p-2 w-full border border-gray-600"
            name="archetype"
          >
            <option
              value="All"
            >
              All Archetypes
            </option>
          </select>
          <select
            class="bg-gray-700/80 text-white rounded-md p-2 w-full border border-gray-600"
            name="artGenGroup"
          >
            <option
              value="ALL_UNITS"
            >
              All Art Gen Groups
            </option>
            <option
              value="NONE"
            >
              None
            </option>
            <option
              value="A"
            >
              Type A
            </option>
            <option
              value="B"
            >
              Type B
            </option>
            <option
              value="C"
            >
              Type C
            </option>
            <option
              value="D"
            >
              Type D
            </option>
            <option
              value="E"
            >
              Type E
            </option>
            <option
              value="EX1"
            >
              Type EX1
            </option>
            <option
              value="EX2"
            >
              Type EX2
            </option>
            <option
              value="EX3"
            >
              Type EX3
            </option>
            <option
              value="F"
            >
              Type F
            </option>
            <option
              value="G"
            >
              Type G
            </option>
            <option
              value="ANY_ART_GEN"
            >
              Any Art Gen Unit
            </option>
          </select>
          <select
            class="bg-gray-700/80 text-white rounded-md p-2 w-full border border-gray-600"
          >
            <option
              value="release"
            >
              Sort: Release Order
            </option>
            <option
              value="A-Z"
            >
              Sort: Name (A-Z)
            </option>
            <option
              value="Z-A"
            >
              Sort: Name (Z-A)
            </option>
          </select>
          <div
            class="flex items-center justify-center p-2 bg-gray-900/50 rounded-md border border-gray-700"
          >
            <label
              class="mr-3 text-sm font-bold text-gray-300 select-none"
              for="my-units-toggle"
            >
              Show My Units
            </label>
            <input
              class="form-checkbox h-5 w-5"
              id="my-units-toggle"
              type="checkbox"
            />
          </div>
        </div>
        <div
          class="mt-4 text-center"
        >
          <button
            class="text-sm font-semibold text-red-400 hover:text-red-300"
          >
            Clear All My Units
          </button>
        </div>
      </div>
      <div
        class="grid grid-cols-4 sm:grid-cols-6 md:grid-cols-8 lg:grid-cols-10 xl:grid-cols-12 gap-x-4 gap-y-8"
      >
        <div
          class="relative group flex flex-col items-center cursor-pointer transform transition-transform duration-200 hover:scale-105"
        >
          <div
            class="absolute inset-0 rounded-lg transition-all duration-300 glow-border"
          />
          <button
            aria-label="Remove from collection"
            class="absolute top-1 right-1 z-10 w-6 h-6 flex items-center justify-center rounded-full text-white font-bold text-sm 
                   transition-all transform group-hover:scale-110 
                   bg-green-500 hover:bg-green-400"
          >
            âœ“
          </button>
          <img
            alt="Hero Alpha Thumbnail"
            class="w-24 h-24 object-cover rounded-lg border-2 border-gray-700/80 group-hover:border-cyan-400 transition-colors shadow-md"
            loading="lazy"
            src="https://placehold.co/128x128/1f2937/9ca3af?text=HER"
          />
          <p
            class="mt-2 text-sm font-semibold text-center text-gray-300 truncate w-24"
          >
            Hero Alpha
          </p>
          <div
            class="absolute bottom-full mb-2 w-48 p-3 bg-gray-900/90 backdrop-blur-sm border border-gray-600 rounded-lg shadow-xl opacity-0 group-hover:opacity-100 transition-all duration-300 pointer-events-none z-10 transform group-hover:-translate-y-1"
          >
            <h4
              class="font-bold text-white text-base"
            >
              Hero Alpha
            </h4>
            <p
              class="text-sm text-gray-400"
            >
              Fire - 
            </p>
            <p
              class="text-sm font-bold text-cyan-400 mt-1"
            >
              Art Gen: None
            </p>
            <div
              class="absolute bottom-[-5px] left-1/2 -translate-x-1/2 w-3 h-3 bg-gray-900/90 border-b border-r border-gray-600 transform rotate-45"
            />
          </div>
        </div>
        <div
          class="relative group flex flex-col items-center cursor-pointer transform transition-transform duration-200 hover:scale-105"
        >
          <div
            class="absolute inset-0 rounded-lg transition-all duration-300 border-2 border-transparent"
          />
          <button
            aria-label="Add to collection"
            class="absolute top-1 right-1 z-10 w-6 h-6 flex items-center justify-center rounded-full text-white font-bold text-sm 
                   transition-all transform group-hover:scale-110 
                   bg-gray-600/80 hover:bg-blue-500"
          >
            +
          </button>
          <img
            alt="Hero Beta Thumbnail"
            class="w-24 h-24 object-cover rounded-lg border-2 border-gray-700/80 group-hover:border-cyan-400 transition-colors shadow-md"
            loading="lazy"
            src="https://placehold.co/128x128/1f2937/9ca3af?text=HER"
          />
          <p
            class="mt-2 text-sm font-semibold text-center text-gray-300 truncate w-24"
          >
            Hero Beta
          </p>
          <div
            class="absolute bottom-full mb-2 w-48 p-3 bg-gray-900/90 backdrop-blur-sm border border-gray-600 rounded-lg shadow-xl opacity-0 group-hover:opacity-100 transition-all duration-300 pointer-events-none z-10 transform group-hover:-translate-y-1"
          >
            <h4
              class="font-bold text-white text-base"
            >
              Hero Beta
            </h4>
            <p
              class="text-sm text-gray-400"
            >
              Water - 
            </p>
            <p
              class="text-sm font-bold text-cyan-400 mt-1"
            >
              Art Gen: None
            </p>
            <div
              class="absolute bottom-[-5px] left-1/2 -translate-x-1/2 w-3 h-3 bg-gray-900/90 border-b border-r border-gray-600 transform rotate-45"
            />
          </div>
        </div>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
